lib/Dialect/ESI/ESIPasses.cpp

```diff
@@ -0,0 +1,106 @@
+//===- ESIToRTL.cpp - ESI to RTL/SV conversion passes -----------*- C++ -*-===//
+//
+// Lower ESI to RTL and SV.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/ESI/ESIDialect.h"
+#include "circt/Dialect/ESI/ESIOps.h"
+#include "circt/Dialect/RTL/Dialect.h"
+#include "circt/Dialect/SV/Dialect.h"
+
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/Support/Debug.h"
+
+#include <memory>
+
+namespace circt {
+namespace esi {
+#define GEN_PASS_CLASSES
+#include "circt/Dialect/ESI/ESIPasses.h.inc"
+} // namespace esi
+} // namespace circt
+
+using namespace mlir;
+using namespace circt::esi;
+
+namespace {
+
+/// Lower `ChannelBuffer`s, breaking out the various options. For now, just
+/// replace with the specified number of pipeline stages (since that's the only
+/// option).
+struct ChannelBufferLowering : public ConversionPattern {
+  ChannelBufferLowering(MLIRContext *ctx)
+      : ConversionPattern(ChannelBuffer::getOperationName(), 1, ctx) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const final {
+    auto loc = op->getLoc();
+    ChannelBuffer buffer = dyn_cast<ChannelBuffer>(op);
+    if (!buffer)
+      return failure();
+
+    ChannelBufferOptions opts = buffer.options();
+    auto type = buffer.getType();
+    Value input = buffer.input();
+
+    // Expand 'abstract' buffer into 'physical' stages.
+    auto stages = opts.stages();
+    size_t numStages = 1;
+    if (stages) {
+      // Guaranteed positive by the parser.
+      numStages = (size_t)stages.getInt();
```

Comment:
nit: I think `getValue` is preferred over `getInt` here.

https://github.com/llvm/llvm-project/blob/c2cb61bed3652126278b4a738e367f524e040ccc/mlir/include/mlir/IR/Attributes.h#L357